sandbox/lua/filters/procstat.lua

```diff
@@ -0,0 +1,200 @@
+    Number of processes in runnable state.  (Linux 2.5.45 onward.)
+
+procs_blocked 2
+    Number of processes blocked waiting for I/O to complete.  (Linux 2.5.45 onward.)
+
+softirq 288977 23 101952 19 13046 19217 7 19125 92077 389 43122
+    Time spent servicing soft-interrupts.
+--]]
+
+-- require "cjson"
+
+
+local l = require 'lpeg'
+require "string"
+
+local proc_stat_previous = nil
+local proc_stat_mappings = {
+    cpu = {
+        'user', 'nice', 'system', 'idle', 'iowait', 'irq',
+        'softirq', 'steal', 'guest', 'guestnice'
+    }
+}
+
+local preserve_data = read_config("preserve_data")
+local blacklist_regex = read_config("blacklist_regex") or 'cpu$'
+
+function process_delta(new, old)
+    -- Recursion - check diff in table or value
+    if type(new) ~= type(old) then return nil end
+    if type(old) ~= "table" then
+        return new - old
+    end
+    local delta = {}
+    for k, v in pairs(old) do
+        delta[k] = process_delta(new[k], v)
+    end
+    return delta
+end
+
+
+function process_fields(delta)
+    -- proccess deltas and calculate cpu average usage on fields of interest
+    local fields = {}
+    for k, v in pairs(delta) do
+        --if type(v) ~= 'table' then
+        --    msg.Fields[k..'_delta'] = v
+        --end
+        if k:match("cpu") ~= nil then
+            local totalticks = 0
+            for ii, vv in pairs(v) do
+                local field_name = proc_stat_mappings['cpu'][ii]
+                fields[k..'_delta_'..field_name] = vv
+                totalticks = totalticks + vv
+            end
+            if totalticks > 0 then
+                -- only pre-calc user, nice, system, idle because all kernels have field 
+                fields[k..'_calc_totalticks'] = totalticks
+                fields[k..'_calc_user'] = fields[k..'_delta_user'] / totalticks
+                fields[k..'_calc_nice'] = fields[k..'_delta_nice'] / totalticks
+                fields[k..'_calc_system'] = fields[k..'_delta_system'] / totalticks
+                fields[k..'_calc_idle'] = fields[k..'_delta_idle'] / totalticks
+                fields[k..'_calc_usage'] = 1 - fields[k..'_calc_idle']
+            end
+        end
+        if k == 'ctext' then
+            fields[k..'_delta'] = v
+        end
+        if k == 'softirq' then
+            for ii, vv in pairs(v) do
+                fields[k..'_delta_'..ii] = vv
+            end
+        end
+    end
+    return fields
+end
+
+
+function process_message()
+    local stats = {}
+    --local basic_stats = { "ctxt", "processes" }
+    local msg = {
+        Type = "procstat",
+        Payload = read_message("Payload"), 
+        Fields = {},
+    }
+    -- read values found in the fields_index into a usable table
+    local wl = whitelist_pattern
+    local data = {}
+    local index_count = read_message("Fields[index_count]")
+    if index_count == nil then return 0 end
+    for i=1, index_count - 1 do
+        f = read_message("Fields[index]", 0, i-1)
```

Comment:
Not sure why the list of fields is being include when they could just be iterated with read_next.  This would simplify a bunch of the following code and also reduce the message size.